### ISSUE TITLE
Automated code analysis

### ISSUE THREAD

**Issue by raimis** [openmm/openmm#3009]
We should consider using automated code analysis tools along the current tests. I have positive experience with https://lgtm.com/, which is free for the open-source projects.

For Python, it works out-of-the-box and can catch many incorrect/not recommended coding patters. For C++, the configuration is a bit more involved.

I'm already trying to set up for https://github.com/openmm/openmm-plumed/ and see how it works.

**Comment by peastman** [comment]
Have you tried running it on OpenMM?  I'd be interested to see what its current report looks like.

**Comment by raimis** [comment]
https://lgtm.com/projects/g/openmm/openmm?mode=list

**Comment by peastman** [comment]
Interesting.  It does find a lot of things that would be good to fix, like unused imports.  But it also complains about a lot of things that are correct.

**Comment by raimis** [comment]
The default behavior could be tweaked with `lgmt.yml`.

**Comment by peastman** [comment]
I'll start by going through the report and fixing anything I think makes sense.

They've made some strange decisions about how to categorize things.  For example, any time you have a loop and don't use the loop variable, such as

```python
            self.atomBonds = [[] for a in self.atoms]
```

it calls that a high severity error???  That's a common and totally ordinary thing to do.  How else would you write that?

**Comment by raimis** [comment]
Use `_` to indicate, the variable is unused intentionally:
```python
self.atomBonds = [[] for _ in self.atoms]
```

**Comment by peastman** [comment]
So they use adherence to a particular set of style guidelines as a proxy for correctness. :(  That's all the more reason it's wrong for them to categorize it as high severity.

**Comment by jchodera** [comment]
> So they use adherence to a particular set of style guidelines as a proxy for correctness. :( That's all the more reason it's wrong for them to categorize it as high severity.

It's hard for even another human to know what you are intending, or if it may have been an algorithmic error, unless you do something specific to indicate that you did not intend to not use the iterable. The `_` tells other people and algorithms "I don't intend to use this."

In general, that's the reason we have style guidelines---to make it easier for other humans to understand what you intend, detect issues, and help out when needed. 

This is not a bad thing---it's a great thing when other humans (and bots) need to help.

**Comment by peastman** [comment]
It's not a bad thing to have style guidelines.  It is a bad thing when an automated error checker misclassifies issues.  Error checkers tend to have a very long tail of things they can report, starting with clear bugs, then suspected bugs, then unusual constructions that might or might not indicate a problem, and so on.  That's why it's critical to categorize them well, so people can spot the definite problems without sifting through hundreds of false positives.

In this case it generated 361 alerts: 15 errors, 163 warnings, and 183 recommendations.  But looking at the 15 items it confidently labeled as "errors", I can see at a glance that most of them actually work correctly.

**Comment by swails** [comment]
I think the reason that style guides tend to get "enforced" to a much higher degree within Python is that the nature of the language means that a very large number of errors that a compiler would normally catch are completely valid, and often "correct", python.

My take is that the severity of issues is assigned based on community-derived heuristics, and that it is very frequently hard to identify incorrect vs. correct uses of certain patterns in Python.  The options boil down to:

1. Warn on everything and annoy everybody
2. Warn on nothing and let lots of issues slip through
3. Adopt conventions -- harmonizing with a style -- to project intent and limit the number of false positives without failing to alert on common errors that are easy for a machine to find, but hard for a human

A very common pattern is to do something like:

```python
for item in items:
    do_something_with(item)
```

using singular and plural this way.  Short of using `mypy` on everything, there's no way for a static analyzer to realize that when you type:

```python
for item in items:
    do_something_with(items)
```

you really meant the first.  While no language with typed names would let such a mistake slip through the compiler, Python does.  Having a static analyzer catch this is incredibly helpful.  But there *are* a lot of times when you'd want to iterate this way just to do something a given number of times (like your example in OpenMM).  Short of building a complex AI static analyzer, you introduce conventions improve the accuracy of the analyzer.  These conventions are rooted in existing parts of the generic Python style guide, but grow into a widely accepted part of the standard proper mainly because these tools become adopted by the community.

> In this case it generated 361 alerts: 15 errors, 163 warnings, and 183 recommendations. But looking at the 15 items it confidently labeled as "errors", I can see at a glance that most of them actually work correctly.

Fair.  But 15 really isn't a lot of issues, and there is most likely a very quick way to satisfy the checker in each case or simply indicate that the checker should ignore it (in Python, most checkers can be disabled on a line with a trailing `# noqa` comment).  Whether or not it's worth doing this comes down to a decision of whether the value provided by static analyzers warrants the work associated with adhering to the conventions of those analyzers.

In my view, the answer is "yes."  This opinion was shared by the majority of developers at both jobs I've worked at so far.  I kind of view it as a cost of picking Python.

**Comment by peastman** [comment]
I'm making my way through the full list of alerts.  So far I haven't found any genuine errors.  There are some things that are worth changing just for general code cleanliness (like the variable names mentioned above), so I'm making those changes.  I've also found some cases where its analysis is simply wrong, such as when it falsely claims that an assignment is unnecessary because the variable gets changed again before it gets used (but actually it doesn't).  But most of them are just worthless warnings about totally correct code.  For example, there are a ton of warnings about lines like

```c++
vector<cl_uint> indexVec(width*numBonds);
```

It complains that we multiply two `int` values, then pass the result to a function that takes a `size_t`, and if the result exceeds the allowed range of an `int` then the cast to `size_t` won't work correctly!

Does anyone have experience with other static checkers for Python or C++?  I'm not impressed with this particular one, but I do think they can be useful in general.  We've been using mypy for DeepChem development.  It can be quite useful, but to get most of the value from it you need to add type annotations, and as I noted on #2548 we're going to be limited in how much we can do that.

**Comment by jchodera** [comment]
I'm with @swails here: These tools exist partly to help enforce standards and conventions, which also makes it easier for others to collaborate on code. 

If there is no performance penalty for doing so, and provided it doesn't consume a huge amount of time, there is no drawback toward aligning the code with standards and conventions that make it easier for others to understand and contribute to the code.

I, for example, would not have known that the multiplication of two `int`s and implicit cast to `size_t` is guaranteed to always be "safe"---especially because there is [conflicting information](http://www.cplusplus.com/forum/general/194034/) that suggests this safety is architecture-independent.

**Comment by peastman** [comment]
Let me emphasize again that I have nothing against static code checkers.  When well written, they can be a useful tool.  It's just that this particular one is badly written.  That's why I asked if anyone had experience with other ones we could try.

> I, for example, would not have known that the multiplication of two ints and implicit cast to size_t is guaranteed to always be "safe"

In this case, we're multiplying a value on the order of 1 (usually either 2 or 4) by a value that is a tiny fraction of the largest representable integer.  There is no risk at all of it overflowing.  That isn't because integer multiplication can't overflow.  It's because we've chosen appropriate types for the values we need to store.

If you really think this is a problem that deserves a warning, then it should warn about every single use of any math operation.  Any addition, subtraction, or multiplication could theoretically overflow.  But that would be silly and useless.  It would produce a huge number of warnings, and only a negligible fraction would correspond to real bugs.  It wouldn't serve the intended function, which is to help you find errors in your code.

So instead they only warn if the result gets cast to a different type.  Presumably someone thought that would make it more useful, but they were wrong.  It's every bit as silly and useless.  We didn't design the vector class.  We didn't choose that its length argument should be a size_t instead of an int.  But that doesn't suddenly make it dangerous to multiply two ints to get the length of a vector.  They still produce a huge number of warnings, of which a negligible fraction are real bugs.  It still is useless for helping you to find errors in your code.

**Comment by jchodera** [comment]
@peastman Did you decide you did _not_ want to have automated code analysis?

**Comment by peastman** [comment]
Not at all.  Static code checkers can be useful when well written.  On DeepChem we're using Mypy which has proven very useful, but I think it would be [much less useful for OpenMM](https://github.com/openmm/openmm/issues/2548#issuecomment-740212426).  Back in my Java programming days we found FindBugs very useful.  The question is whether we can find a checker for C++ and/or Python that provides enough benefit to justify the effort.
